# The impact of exercise on gene regulation in association with complex trait genetics

This is the GitHub repository associated with the publication: [insert publication here] 

It primarily hosts scripts written in the R programming language that were used to 
generate all of the paper figures. The names of these scripts are given in figure order, 
i.e. `fig*_`. Software and data dependencies are listed below. Data dependencies are 
also provided in the `/data` folder where possible.

Additionally, this GitHub plays host to several supplementary figures, datasets, 
MCMC output, and other files generated by these scripts.

In the interests of reproducibility, I've endeavored that, assuming all dependencies 
are properly installed, these scripts function as end-to-end pipelines. That said, 
just because they ran on my machine (macOS 10.15.7) does not mean they'll run seamlessly 
on yours, at least if you're reading this before I've had the chance to dockerize everything. 
Please contact me at nikgvetr@stanford.edu if you encounter difficulties and I'll do me best to help.

## Dependencies

### R packages 

* Cairo  
* DESeq2  
* EnsDb.Hsapiens.v79  
* EnsDb.Hsapiens.v86  
* MASS  
* MotrpacRatTraining6mo  
* MotrpacRatTraining6moData  
* arrow  
* biomaRt  
* caret  
* circlize  
* clusterProfiler  
* cmdstanr  
* data.table  
* doParallel  
* dplyr  
* edgeR  
* ensembldb  
* fitdistrplus  
* foreach  
* ggplot2  
* invgamma  
* jpeg  
* jsonlite  
* ks  
* limma  
* org.Hs.eg.db  
* org.Rn.eg.db  
* parallel  
* plotrix  
* posterior  
* pracma  
* sparklyr  
* sparklyr.nested  
* testit  
* xlsx  

### Command-line tools

* [plink2](https://www.cog-genomics.org/plink/2.0/)  
* [GCTA](https://yanglab.westlake.edu.cn/software/gcta/#Overview)  
* [MESC](https://github.com/douglasyao/mesc)  
* [LDSC](https://github.com/bulik/ldsc)  
* [Stan](https://mc-stan.org/cmdstanr/)  

### Datasets  

The following paths, files, and R data objects are used within these scripts. Here
we provide documentation on their sources. When possible, these datasets are also
included in `/data`. 

* `/Volumes/SSD500GB/gtex-pipeline/expression_genotypes/GTEx_v8_noChr_noRSID.bim`:  
* `/Volumes/SSD500GB/gtex-pipeline/GTEx_Analysis_v8_eQTL_covariates/`:  
* `/Volumes/SSD500GB/gtex-pipeline/GTEx_Analysis_v8_eQTL_expression_matrices/`:  
* `/Volumes/SSD500GB/gtex-pipeline/log2-normalized-expression/log2-normalized-expression_*.expression.bed.gz`:  
* `~/data/smontgom/41467_2021_23579_MOESM6_ESM.csv`: [Dataset 5](https://static-content.springer.com/esm/art%3A10.1038%2Fs41467-021-23579-x/MediaObjects/41467_2021_23579_MOESM6_ESM.xlsx)
from the human exercise meta-analysis by Amar et al.: <https://pubmed.ncbi.nlm.nih.gov/34108459/> 
* `~/data/smontgom/eqtl/spredixcan_igwas_gtexmashrv8_*`: Barbeira et al. `S-PrediXcan` results
available [here](https://zenodo.org/record/3518299#.Y9rPqezMIUE), provided as a 
supplement to their manuscript: <https://pubmed.ncbi.nlm.nih.gov/33499903/>   
* `~/data/smontgom/est_gcor_mat.RData`:  
* `~/data/smontgom/GENES_HUMAN.txt`:  
* `~/data/smontgom/GENES_RAT.txt`:  
* `~/data/smontgom/GTEx_Analysis_v8_sbgenes/signif.sbgenes.txt`:  
* `~/data/smontgom/GTEx_v8_ExpressionScores/tissues/`:  
* `~/data/smontgom/gwas_metadata.csv`:  
* `~/data/smontgom/imputed_gwas_hg38_1.1/`: GWAS imputed and harmonized for colocalization 
analysis with GTEx v8. Provided [here](https://zenodo.org/record/3629742#.Y9rTQOzMIUF) by Barbeira et al. as a supplement to their 
manuscript: <https://pubmed.ncbi.nlm.nih.gov/33499903/>  
* `~/data/smontgom/meta_analysis_results.RData`:  
* `~/data/smontgom/old_dea_deseq_20201121/*_training-dea_20201121.RData`:  
* `~/data/smontgom/open-targets_tissue-x-disease_*`:  
* `~/data/smontgom/opentargets/associationByOverallDirect.csv`:  
* `~/data/smontgom/opentargets/associationByOverallIndirect.csv`:  
* `~/data/smontgom/PANTHER17_human_rat_ref_genome_orthologs.tsv`:  
* `~/data/smontgom/relative_expression_motrpac_gtex`:  
* `~/data/smontgom/RGD_ORTHOLOGS_20201001.txt`: Data downloaded from RGD FTP website 
on 10/01/2020 by Pierre Jean (<ftp://ftp.rgd.mcw.edu/pub/data_release/orthologs/>). 
This file contain mapping between rat, human, and mouse primarily through NCBI gene IDs.  
* `~/data/smontgom/RSID_POS_MAP_*.txt`:  
* `~/data/smontgom/zcor_transcriptome_pass1b.tsv`:  
* `~/repos/ldsc/1000G_EUR_Phase3_plink/1000G.EUR.QC.*.bim`:  
* `~/repos/ldsc/baseline/baseline.*.annot`:  
* `~/repos/ldsc/custom_genesets/cluster_*.chr_*.annot`: 
* `~/repos/ldsc/ENSG_coord.txt`:  
